android/PhysicalWeb/app/src/main/java/org/physical_web/physicalweb/Utils.java

```diff
@@ -91,48 +89,103 @@ private static int getCurrentPwsEndpointVersion(Context context){
     return Integer.parseInt(endpoint.split(",")[1]);
   }
 
-  public static String convertToSettingsEndpointFormat(String pwsUrl, int pwsVerion){
-    return pwsUrl + "," + pwsVerion;
+  private static String getCurrentPwsEndpointApiKey(Context context){
+    String endpoint = getCurrentPwsEndpoint(context);
+    if (endpoint.endsWith(",") || endpoint.endsWith("null")){
+      return "";
+    }
+    return endpoint.split(",")[2];
+  }
+
+  private static String readCustomPwsEndpointUrl(Context context){
+    SharedPreferences sharedPref = PreferenceManager.getDefaultSharedPreferences(context);
+    return sharedPref.getString(context.getString(R.string.custom_pws_url_key), "");
+  }
+
+  private static int readCustomPwsEndpointVersion(Context context){
+    SharedPreferences sharedPref = PreferenceManager.getDefaultSharedPreferences(context);
+    return Integer.parseInt(sharedPref.getString(
+      context.getString(R.string.custom_pws_version_key),
+      context.getString(R.string.custom_pws_version_default)));
+  }
+
+  private static String readCustomPwsEndpointApiKey(Context context){
+    SharedPreferences sharedPref = PreferenceManager.getDefaultSharedPreferences(context);
+    return sharedPref.getString(context.getString(R.string.custom_pws_api_key_key), "");
+  }
+
+  public static String convertToSettingsEndpointFormat(String pwsUrl, int pwsVerion,
+                                                       String apiKey){
+    return pwsUrl + "," + pwsVerion + "," + apiKey;
```

Comment:
You could use the NUL character (or any other non-printable char) since it should never appear in a valid URL.  Java strings aren't NUL-terminated.

Also I just noticed pwsVerion is missing an S.